soccer/Referee.cpp

```diff
@@ -110,154 +29,186 @@ static const bool CancelBallPlaceOnHalt = true;
+
 void Referee::spinKickWatcher(const SystemState& system_state) {
-    if (system_state.ball.valid) {
-        /// Only run the kick detector when the ball is visible
-        switch (_kickDetectState) {
-            case WaitForReady:
-                /// Never kicked and not ready for a restart
-                break;
-
-            case CapturePosition:
-                // Do this in the processing thread
-                _readyBallPos = system_state.ball.pos;
+    /// Only run the kick detector when the ball is visible
+    if (!system_state.ball.valid) {
+        return;
+    }
+
+    switch (_kickDetectState) {
+        case WaitForReady:
+            /// Never kicked and not ready for a restart
+            break;
+
+        case CapturePosition:
+            // Do this in the processing thread
+            _readyBallPos = system_state.ball.pos;
+            _kickDetectState = WaitForKick;
+            break;
+
+        case WaitForKick:
+            if (system_state.ball.pos.nearPoint(_readyBallPos, KickThreshold)) {
+                return;
+            }
+            // The ball appears to have moved
+            _kickTime = RJ::Time();
+            _kickDetectState = VerifyKick;
+            break;
+
+        case VerifyKick:
+            if (system_state.ball.pos.nearPoint(_readyBallPos, KickThreshold)) {
+                // The ball is back where it was.  There was probably a
+                // vision error.
                 _kickDetectState = WaitForKick;
-                break;
-
-            case WaitForKick:
-                if (!system_state.ball.pos.nearPoint(_readyBallPos,
-                                                     KickThreshold)) {
-                    // The ball appears to have moved
-                    _kickTime = QTime::currentTime();
-                    _kickDetectState = VerifyKick;
-                }
-                break;
-
-            case VerifyKick:
-                if (system_state.ball.pos.nearPoint(_readyBallPos,
-                                                    KickThreshold)) {
-                    // The ball is back where it was.  There was probably a
-                    // vision error.
-                    _kickDetectState = WaitForKick;
-                } else if (_kickTime.msecsTo(QTime::currentTime()) >=
-                           KickVerifyTime_ms) {
-                    // The ball has been far enough away for enough time, so
-                    // call this a kick.
-                    _kickDetectState = Kicked;
-                }
-                break;
-
-            case Kicked:
-                // Stay here until the referee puts us back in Ready
-                break;
-        }
+            } else if (auto ms_elapsed = std::chrono::duration_cast<
+                                             std::chrono::milliseconds>(
+                                             RJ::Time() - _kickTime)
+                                             .count();
```

Comment:
This is a lot in an if statement, it might be better to pull it out in front, especially since this is already split up

Reply:
ok